framework/test/mocha/integration/delegates/forge.js

```diff
@@ -0,0 +1,113 @@
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const util = require('util');
+const { transfer } = require('@liskhq/lisk-transactions');
+const application = require('../../common/application');
+const Bignum = require('../../../../src/modules/chain/helpers/bignum');
+const random = require('../../common/utils/random');
+const accountFixtures = require('../../fixtures/accounts');
+
+const genesisBlock = __testContext.config.genesisBlock;
+const { NORMALIZER } = global.__testContext.config;
+
+const localCommon = require('../common');
+
+const addTransactionsAndForge = util.promisify(
+	localCommon.addTransactionsAndForge
+);
+const transactionInPool = localCommon.transactionInPool;
+const addTransaction = util.promisify(localCommon.addTransaction);
+const fillPool = util.promisify(localCommon.fillPool);
+const forge = util.promisify(localCommon.forge);
+
+let library;
+
+describe('delegates (forge)', () => {
+	before(done => {
+		application.init(
+			{
+				sandbox: {
+					name: 'lisk_test_delegates_forge',
+				},
+			},
+			(err, scope) => {
+				library = scope;
+				library.modules.blocks.lastBlock.set(genesisBlock);
+				done();
+			}
+		);
+	});
+
+	after(done => {
+		application.cleanup(done);
+	});
+
+	describe('forge', () => {
+		describe('total spending', () => {
+			it('should not include transactions which exceed total spending per account balance', async () => {
+				// Transfer 1LSK to account
+				const account = random.account();
+				const transaction = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(1).toString(),
+					recipientId: account.address,
+					passphrase: accountFixtures.genesis.passphrase,
+				});
+				await addTransactionsAndForge(library, [transaction], 0);
+
+				// Transfer 0.5 LSK to a random account - The account balance would be 0.4
+				const trs_06 = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(0.6).toString(),
+					recipientId: random.account().address,
+					passphrase: account.passphrase,
+				});
+
+				// Transfer 0.5 LSK to a random account - The account balance would be -0.2
+				const trs_04 = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(0.4).toString(),
+					recipientId: random.account().address,
+					passphrase: account.passphrase,
+				});
+
+				// Transfer 0.3 LSK to that account - The account balance would be 0.1
+				const trs_03 = transfer({
+					amount: new Bignum(NORMALIZER).multipliedBy(0.3).toString(),
+					recipientId: account.address,
+					passphrase: accountFixtures.genesis.passphrase,
+				});
+
+				// Add the transactions to pool and forge
+				await addTransaction(library, trs_06);
```

Comment:
This is the happy path. The order should be `credit03`, `spend06`, `spend04`. Transaction pool should not select those transactions to be included in one block, because the block will be invalid when it hit block processing (`_verifyTotalSpending`). I don't see any changes related to transaction pool in this PR, so I assume the test with such an order will fail and block with all 3 transactions will be created - which is not an expected behavior.

Reply:
I tested with the correct order and it forge a block with all 3 transactions and fail to process it:
```
	Add transaction ID: 11584432913339092061
	Add transaction ID: 12970355266971244115
	Add transaction ID: 15590495324329756252
		Last block height: 2 Last block ID: 16453677599573165466 Last block timestamp: 10 Next slot: 2 Next delegate PK: 74583aba9c0b92e4f08c8c75e6df341c255ca007971195ff64d6f909dc4b7177 Next block timestamp: 20
Error: the array [
  "Transaction: 12970355266971244115 failed at .amount: Account does not have enough LSK for total spending. balance: 100000000, spending: 120000000"
] was thrown, throw an Error :)
```

Reply:
Let me add more tests and confirm the behaviour.

Reply:
@4miners In any order you pass, the transactions are validated individually in the pool. And then the block is generated with sorted transactions and invalid transition is removed.

Reply:
@nazarhussain Invalid transaction is not removed. It forges a block which contains all of the transactions, then it passes it to block processing and it fails there, so the `forge` call fails as the error propagates.

Reply:
This is what I saw during the debugging and while calling the `fillPool`;

```
%% before fill pool 

{ received: Queue { _transactions: [], _index: {} },
  validated: Queue { _transactions: [], _index: {} },
  verified:
   Queue {
     _transactions:
      [ [TransferTransaction],
        [TransferTransaction],
        [TransferTransaction] ],
     _index:
      { '2051428277466744290': [TransferTransaction],
        '17229733069888235767': [TransferTransaction],
        '4415635931455692879': [TransferTransaction] } },


%% after fill pool 

{ received: Queue { _transactions: [], _index: {} },
  validated: Queue { _transactions: [], _index: {} },
  verified:
   Queue {
     _transactions: [ [TransferTransaction], [TransferTransaction] ],
     _index:
      { '2051428277466744290': [TransferTransaction],
        '4415635931455692879': [TransferTransaction] } },
  pending: Queue { _transactions: [], _index: {} },
  ready: Queue { _transactions: [], _index: {} } }
```

From this debug information I tend to think it it removes the invalid transaction from the pool and keep rest of two in the `verified` queue.

Reply:
Please change the order of the transactions to `credit03`, `spend06`, `spend04` and re-run this test - it will fail to forge.